chore: recategorize apps

rjackson opened the pull request:
## What does this PR do?

This PR updates the `config.json` or `_metadata.ts` files of our app store apps, to recategorise them as per the Google Sheet link provided by in #9304.

There are also a number of supporting changes to ensure the app store & categories continue to work correctly after the recategorisation:

- Update the static list of categories (getAppCategories.ts) as per the desired set of categories
- Update the AppCategories enum, so apps can be saved in the database as per the new categories
- Replaced some static categories lists with the enum, to improve type-safety and consistency between the code & database
- Fixed the "Installed Apps" and "Admin Apps" lists improperly handling new or changed categories
  - The lists sometimes served a 404 for the new categories, due to the use of static category lists
  - The admin apps list had a bug where it didn't always show the "There are no apps in this category" messaging
- Updated the `seed-app-store.ts` file, so the set of apps that come alongside a dev environment are in-sync with the app metadata
- Replaced `seed-app-store.config.json` with loading each app's config / metadata entries directly, so its guaranteed to be in-sync with the apps we have in the codebase
- Added `syncAppMeta` cron to keep the App metadata stored in the database (categories, dirName) in sync with the raw config values.
- Added redirects for "video" category pages to "conferencing" pages

**Differences from the spreadsheet**

There were a few apps in the spreadsheet that were missing from the codebase, which I've therefore not recategorised (because there's nothing to recategorise):

- msteams (this was actually office365-video, which has been updated)
- slack
- space-booking

And a few apps that were missing from the spreadsheet, which I've recategorised as:

App | Before | After
-- | -- | --
Discord | Video | conferencing
eightxeight | video | conferencing
mirotalk | video | conferencing
sylapsvideo | video | conferencing
event-type-location-static | video | conferencing
webex | video | conferencing
zohocrm | other | crm
metapixel | analytics | (no change)
office365-video | conferencing | (no change)
rainbow | web3 | (no change)
basic | other | (no change)
booking-pages-tag | analytics | (no change)
event-type-app-card | other | (no change)
general-app-settings | other | (no change)
link-as-an-app | other | (no change)
zoho-bigin | other | (no change)

<strike>
**Potentially outstanding**

This PR does not include any data migrations to correct apps that are already in the database's `App` table.

This does mean that the old categories might still be visible and accessible from the "Apps Store" pages, as those pages infer the supported categories from the apps in the database, rather than using the static list from `getAppsCategories` or the `AppCategories` enum.

Some options we could explore here:

1. Write a data migration to update the database entries
2. Refactor the Apps Store pages to use the `getAppCategories` and load via app-store metadata, rather than what's in the database
3. Or, a hybrid of both approaches.

I don't think this is an immediate issue, because inferring the valid categories from database entries ensures forward/backward compatibility as we recategorise apps.
</strike>

Fixes #9304

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

**Environment**: Staging(main branch) / Production

## Type of change


- Bug fix (non-breaking change which fixes an issue)
- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

| Page            | See the new categories? | See apps in the right category? |
| --------------- | ----------------------- | ------------------------------- |
| Apps Store      | Yes                     | Yes                             |
| Installed Apps  | Yes                     | Yes                             |
| Categories list | Yes                     | Yes                             |
| Category page | (n/a) | Yes |
| Admin apps list | Yes                     | Yes                             |

Page | Screenshot
--- | ---
App Store | <img width="1732" alt="app store" src="https://github.com/calcom/cal.com/assets/602850/2a66522f-e7db-4d32-94d6-7dd407a43777">
Installed apps | <img width="1732" alt="installed apps 2" src="https://github.com/calcom/cal.com/assets/602850/5e00ad38-1193-4c63-953c-19acd4ba9a9b">
Categories list | <img width="1732" alt="categories list" src="https://github.com/calcom/cal.com/assets/602850/66b33caa-0214-4137-b8de-c44e2f2f74ce">
Category Page | <img width="1732" alt="category page" src="https://github.com/calcom/cal.com/assets/602850/9a6466ef-17f1-4516-a04d-b2e1712ea1ce">
Admin apps list | <img width="1732" alt="admin apps 2" src="https://github.com/calcom/cal.com/assets/602850/e6d2ce4f-016a-4f7d-a751-544e0200a5d3">

## Checklist

- [x] Ensure no breakages in tests (currently waiting for e2e tests to finish, they're taking a while!)
- [x] Add `apps/web/public/app-categorise/messaging.svg` (Job for you @PeerRich?)
- [x] If we do a data migration: Redirect `/apps/categories/video` to `/apps/categories/conferencing`?


Solves #9304
/claim #9304

rjackson committed:
Remove unused code in InstalledAppsLayout

rjackson committed:
Add new app categories "crm", "conferencing" and "messaging"

rjackson committed:
Sort getAppCategories entries alphabetically

rjackson committed:
Fix 404s on new category pages (and remove hardcoded category lists)

rjackson committed:
Fix admin apps list not showing "no available apps" for new categories

rjackson committed:
Recategorise apps

rjackson committed:
Sync seed-app-store categories with config files

rjackson committed:
Replace unnecessary seed-app-store.config.json with appStoreMetadata

rjackson committed:
Copy video.svg to conferencing.svg

PeerRich committed:
Merge branch 'main' into recategorise-apps

PeerRich commented:
amazing PR + description! thank you so much. super quick too

PeerRich commented:
@hariombalhara can you look into this:

> **Potentially outstanding**
> 
> This PR does not include any data migrations to correct apps that are already in the database's App table.
> 
> This does mean that the old categories might still be visible and accessible from the "Apps Store" pages, as those pages infer the supported categories from the apps in the database, rather than using the static list from getAppsCategories or the AppCategories enum.
> 
> Some options we could explore here:
> 
> Write a data migration to update the database entries
> Refactor the Apps Store pages to use the getAppCategories and load via app-store metadata, rather than what's in the database
> Or, a hybrid of both approaches.
> I don't think this is an immediate issue, because inferring the valid categories from database entries ensures forward/backward compatibility as we recategorise apps.

PeerRich commented:
@rjackson can you remove Web3? Huddle can be moved into conferencing

PeerRich commented:
![messaging](https://github.com/calcom/cal.com/assets/8019099/0d1ed017-d8b7-4f30-bb35-01959526c70e)

here is the messaging.svg

PeerRich commented:
i saw you went above and beyond, adding another bounty!

PeerRich commented:
/tip 100

PeerRich commented:
/tip

rjackson committed:
Add messaging.svg

rjackson committed:
Remove web3 from getAppCategories (used by installed apps, admin apps)

rjackson committed:
Fix app-store-cli categories

- Add conferencing
- Add CRM
- Remove video
- Remove web3

rjackson committed:
Remove outdated web3 comment in seed-app-store

rjackson commented:
Thanks @PeerRich 

I've updated the PR with:

- Add messaging.svg
- Remove web3 from static list of categories (getAppCategories), so it will no longer appear in the "Installed Apps" or "Admin -> Apps" lists
  - Web3 apps may still appear if users ever browse to `/apps/categories/web3` directly, or the installation they're using still has apps in their database with the category `web3` (the "potentially outstanding" note you quoted above)
  - This PR still has Rainbow as a web3 app, but that is due to be removed separately via #9299
- Fixed `yarn app-store create` having an outdated list of categories

That last one being another case of having various different lists of categories spread throughout the codebase. I've added some TODO comments suggesting we refactor those later to all pull from a centralised list (refactor getAppCategories to be that centralised definition?). I'd be happy to look into that one separately; I think it'd be easier to test & review as an isolated PR rather than cobbled on the end of this already (slightly)-larger-than-expected PR.

Re: Huddle, that's now categorised as `conferencing` after these changes. I've found & removed an outdated comment suggesting its web3 related, if that was what you picked up on? Let me know if there's something else there I've missed?

I'll have the PR screenshots updated too. The "Installed Apps" and "Admin apps list" will have changed with the removal of "Web3".

PeerRich commented:
amazing work! we will test and likely merge Monday!

alannnc committed:
Merge branch 'main' into recategorise-apps

a0m0rajab commented:
Just a quick thought, if the categories changed again in the future such hand work would be needed to solve it, is there a away to get the categories and apps from the database? or even auto change it with CLI or sometihng?

hariombalhara commented on packages/prisma/seed-app-store.config.json:
@rjackson What's the reason to delete this file?

rjackson replied:
The data in here was essentially a duplicate of `packages/app-store/<app>/config.json` or `.../_metadata.ts`, and it was only used by `seed-app-store.ts`, loaded when setting up dev environments.

This ended up causing me a bit of confusion, because after recategorising all of the app's config or metadata files the `/apps/` and `/apps/categories` pages were still listing the old categories for me.

As we already had the underlying data accessible via code (`@calcom/app-store/appStoreMetaData`), I thought this duplication was not necessary and removed it in favour of using the metadata from the config or metadata files directly.

Relevant commit: 37b685b1d664f7db6fd15e02124687c4b80f649d

rjackson committed:
Merge remote-tracking branch 'origin/main' into recategorise-apps

rjackson committed:
Merge branch 'main' into recategorise-apps

hariombalhara commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1265,6 +1269,8 @@
   "connect_other_apps": "Connect other apps",
   "connect_web3_apps": "Connect web3 apps",
+  "connect_messaging_apps": "Connect messaging apps",
+  "connect_crm_apps": "Connect crm apps",
```
```suggestion
  "connect_crm_apps": "Connect CRM apps",
```

hariombalhara committed:
Update apps/web/public/static/locales/en/common.json

hariombalhara reviewed (changes requested):
Thanks @rjackson 🙏 Looks good mostly.
Have few concerns, before I merge this.
- Prisma migration has a warning that on PostgreSQL 11 or less, the migration might fail. So, need confirmation on that from @zomars @keithwillcode. I think it would be best to break the migration into 2 files as some of the self hosters(if not us) might be using a lower version
- There is no migration to change existing apps from `video` to `conferencing` and other similar category changes. Without the migration we would still see those apps under Video category in /apps page. This is because there is some duplication here(to keep apps listing page fast I guess), the categories from config.json are picked and stored in DB. 
<img width="195" alt="Screenshot 2023-06-12 at 2 06 44 PM" src="https://github.com/calcom/cal.com/assets/1780212/19593e01-e40b-4484-9481-efb3af797a2d">

- There are missing translation strings for installed_app_CATEGORY_description for new categories
<img width="1459" alt="Screenshot 2023-06-12 at 2 10 17 PM" src="https://github.com/calcom/cal.com/assets/1780212/d2abd688-1e93-4d19-8f9a-0cddc97cc3f7">

hariombalhara commented:
> Am I correct in reading this as separate files for:
>
> Changes to AppCategories Enum
> Data correction for apps in database

I think you might have to [move these queries](https://github.com/calcom/cal.com/pull/9306/files/277234d819d7ebda5c679328883ba3a0be941cce#diff-b7985775d1edd2c527071bc4c9d46b1a38ca1a007f5c981cf12ecdb2d33c8201R9) to separate files. 


> Yeah I was curious how we wanted to handle that: Whether there's some process y'all already have to re-sync app metadata & the database, or we look into a migration to update the database, or some refactors to remove this duplication of data (and ensure that app metadata is the source of truth?)
Unfortunately, there is no such process right now to keep them in sync. A process likes that sounds good, because we can't remove the duplicates. Duplicate in DB is to speed up the app-store listing page which has to query the categories of all the apps(vs reading those from files). Although, I think we are doing SSG for app listing page as of now, so that shouldn't matter. @zomars might have more inputs here as he did the initial work on App Store.

> Ah darn, the "Conferencing" category already appeared to be in-place when I made my branch so I assumed its strings were all set up already – I can see the issue you've caught on production, for example 😆
lol 😅

rjackson committed:
Add cron script to keep db apps in sync with app metadata

rjackson committed:
Add redirect for app category "video" to "conferencing"

keithwillcode commented on packages/prisma/migrations/20230603115613_reorganise_app_categories/migration.sql:
```diff
@@ -0,0 +1,10 @@
+-- AlterEnum
+-- This migration adds more than one value to an enum.
+-- With PostgreSQL versions 11 and earlier, this is not possible
```
@hariombalhara 13.7

rjackson committed:
Fix up "video" category overrides to apply to conferencing

rjackson committed:
Fix conferencing apps not showing as a location for non-team users

rjackson commented:
Noting in case its useful for anybody else: The following localisation strings are required for app categories:

```
(category)
installed_app_(category)_headline
installed_app_(category)_description
no_category_apps_description_(category)
connect_(category)_apps
```

rjackson committed:
Restore "installed_app" string for conferencing apps

rjackson commented:
@hariombalhara I've updated the PR as per your feedback.

Prisma doesn't really offer a nice way of running scripts for data migrations, and I didn't really fancy writing and hardcoding all the _current_ categories in the SQL migration. Instead I've gone for something a little more future-proof: A cron script that syncs app entries in the database with the app metadata files, and a Github Action to run that on the 1st of each month.

This sync script can also be triggered manually with the same curl request used by the Github Action:

https://github.com/calcom/cal.com/blob/97f2dc766cd21abcd590538dea148eeea508ab3c/.github/workflows/cron-syncAppMeta.yml#L20-L24

Substituting in appropriate values, of course.

Example output (applied to a db seeded on `main`):

```
@calcom/web:dev: 🧐 Checking DB apps are in-sync with app metadata
@calcom/web:dev: 🔨 Updating app daily-video with categories
@calcom/web:dev: ✅ App gtm is up-to-date and correct
@calcom/web:dev: ✅ App apple-calendar is up-to-date and correct
@calcom/web:dev: ✅ App fathom is up-to-date and correct
@calcom/web:dev: ✅ App caldav-calendar is up-to-date and correct
@calcom/web:dev: 🔨 Updating app jitsi with categories
@calcom/web:dev: ✅ App plausible is up-to-date and correct
@calcom/web:dev: 🔨 Updating app wipe-my-cal with categories
@calcom/web:dev: 🔨 Updating app discord with categories
@calcom/web:dev: 🔨 Updating app huddle01 with categories
@calcom/web:dev: ✅ App wordpress is up-to-date and correct
@calcom/web:dev: 🔨 Updating app routing-forms with categories
@calcom/web:dev: 🔨 Updating app whereby with categories
@calcom/web:dev: ✅ App ga4 is up-to-date and correct
@calcom/web:dev: 🔨 Updating app around with categories
@calcom/web:dev: 🔨 Updating app sylapsvideo with categories
@calcom/web:dev: 🔨 Updating app riverside with categories
@calcom/web:dev: ✅ App pipedream is up-to-date and correct
@calcom/web:dev: 🔨 Updating app typeform with categories
@calcom/web:dev: 🔨 Updating app ping with categories
@calcom/web:dev: 🔨 Updating app sirius_video with categories
@calcom/web:dev: 🔨 Updating app campfire with categories
@calcom/web:dev: 🔨 Updating app raycast with categories
@calcom/web:dev: 🔨 Updating app sendgrid with categories
@calcom/web:dev: ✅ App n8n is up-to-date and correct
@calcom/web:dev: ✅ App exchange is up-to-date and correct
@calcom/web:dev: 🔨 Updating app closecom with categories
@calcom/web:dev: ✅ App qr_code is up-to-date and correct
@calcom/web:dev: ✅ App weather_in_your_calendar is up-to-date and correct
@calcom/web:dev: 🔨 Updating app whatsapp with categories
@calcom/web:dev: 🔨 Updating app telegram with categories
@calcom/web:dev: 🔨 Updating app signal with categories
@calcom/web:dev: ✅ App vimcal is up-to-date and correct
@calcom/web:dev: ✅ App amie is up-to-date and correct
@calcom/web:dev: 🔨 Updating app facetime with categories
@calcom/web:dev: 🔨 Updating app zohocrm with categories
@calcom/web:dev: 🔨 Updating app webex with categories
@calcom/web:dev: ✅ App cron is up-to-date and correct
```

As this solves the lasting database-driven categories, I've also included redirects to forward any visits to the "video" category to the "conferencing" category.

rjackson committed:
Make linter happier

rjackson committed:
Merge remote-tracking branch 'origin/main' into recategorise-apps

rjackson committed:
Remove my "installed_app_conferencing_description" as this was fixed upstream

rjackson committed:
Merge remote-tracking branch 'origin/main' into recategorise-apps

ericrommel commented:
Minor finding, a misspelling here:
![image](https://github.com/calcom/cal.com/assets/10250819/e223b769-2857-42a4-b487-490b301a1a63)

zomars commented on apps/web/components/eventtype/EventSetupTab.tsx:
```diff
@@ -117,8 +117,17 @@ export const EventSetupTab = (
-  const locationOptions = props.locationOptions.filter((option) => {
-    return !team ? option.label !== "Conferencing" : true;
+  const locationOptions = props.locationOptions.map((locationOption) => {
+    console.log(locationOption.options);
```
```suggestion
```

zomars commented on apps/web/pages/api/cron/syncAppMeta.ts:
```diff
@@ -0,0 +1,58 @@
+
+/**
+ * syncAppMeta makes sure any app metadata that has been replicated into the database
+ * remains synchronised with any changes made to the app config files.
```
```suggestion
 * remains synchronized with any changes made to the app config files.
```

zomars reviewed (dismissed):
Codewise LGTM. Delegating tests to QA 🙏🏽

zomars committed:
Merge branch 'main' into recategorise-apps

hariombalhara commented:
> @hariombalhara I've updated the PR as per your feedback.
> 
> Prisma doesn't really offer a nice way of running scripts for data migrations, and I didn't really fancy writing and hardcoding all the _current_ categories in the SQL migration. Instead I've gone for something a little more future-proof: A cron script that syncs app entries in the database with the app metadata files, and a Github Action to run that on the 1st of each month.
> 
> This sync script can also be triggered manually with the same curl request used by the Github Action:
> 
> https://github.com/calcom/cal.com/blob/97f2dc766cd21abcd590538dea148eeea508ab3c/.github/workflows/cron-syncAppMeta.yml#L20-L24
> 
> Substituting in appropriate values, of course.
> 
> Example output (applied to a db seeded on `main`):
> 
> ```
> @calcom/web:dev: 🧐 Checking DB apps are in-sync with app metadata
> @calcom/web:dev: 🔨 Updating app daily-video with categories
> @calcom/web:dev: ✅ App gtm is up-to-date and correct
> @calcom/web:dev: ✅ App apple-calendar is up-to-date and correct
> @calcom/web:dev: ✅ App fathom is up-to-date and correct
> @calcom/web:dev: ✅ App caldav-calendar is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app jitsi with categories
> @calcom/web:dev: ✅ App plausible is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app wipe-my-cal with categories
> @calcom/web:dev: 🔨 Updating app discord with categories
> @calcom/web:dev: 🔨 Updating app huddle01 with categories
> @calcom/web:dev: ✅ App wordpress is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app routing-forms with categories
> @calcom/web:dev: 🔨 Updating app whereby with categories
> @calcom/web:dev: ✅ App ga4 is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app around with categories
> @calcom/web:dev: 🔨 Updating app sylapsvideo with categories
> @calcom/web:dev: 🔨 Updating app riverside with categories
> @calcom/web:dev: ✅ App pipedream is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app typeform with categories
> @calcom/web:dev: 🔨 Updating app ping with categories
> @calcom/web:dev: 🔨 Updating app sirius_video with categories
> @calcom/web:dev: 🔨 Updating app campfire with categories
> @calcom/web:dev: 🔨 Updating app raycast with categories
> @calcom/web:dev: 🔨 Updating app sendgrid with categories
> @calcom/web:dev: ✅ App n8n is up-to-date and correct
> @calcom/web:dev: ✅ App exchange is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app closecom with categories
> @calcom/web:dev: ✅ App qr_code is up-to-date and correct
> @calcom/web:dev: ✅ App weather_in_your_calendar is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app whatsapp with categories
> @calcom/web:dev: 🔨 Updating app telegram with categories
> @calcom/web:dev: 🔨 Updating app signal with categories
> @calcom/web:dev: ✅ App vimcal is up-to-date and correct
> @calcom/web:dev: ✅ App amie is up-to-date and correct
> @calcom/web:dev: 🔨 Updating app facetime with categories
> @calcom/web:dev: 🔨 Updating app zohocrm with categories
> @calcom/web:dev: 🔨 Updating app webex with categories
> @calcom/web:dev: ✅ App cron is up-to-date and correct
> ```
> 
> As this solves the lasting database-driven categories, I've also included redirects to forward any visits to the "video" category to the "conferencing" category.

That's so cool @rjackson. Thank you 🙏

hariombalhara commented on apps/web/pages/api/cron/syncAppMeta.ts:
```diff
@@ -0,0 +1,58 @@
+      dbApp.categories.length !== app.categories.length ||
+      !dbApp.categories.every((category) => app.categories.includes(category))
+    ) {
+      updates["categories"] = app.categories as AppCategories[];
```
It is updated only when there is a category for an App in DB that doesn't exist now.
We should also do the reverse thing, where there is an category in file for an app but it doesn't exist in DB.

rjackson replied:
I think I understand what you're referring to here, and I think that's already happening – I could have dropped a comment to make it a little clearer what these conditions are doing:

By combining the length check and "every" check, these two conditions are making sure the two lists are identical in content. Let me re-validate my logic by typing out the various scenarios:

1. If we have a category removed in the app metadata, then `dbApp.categories.length` will be greater than `app.categories.length`. The first condition will therefore pass, and the categories will be updated.

2. If we have a new category added in the app metadata (the case you're highlighting), the same will be true – just in the other direction.

3. If we had both a category removal and addition then the length of those lists will be the same. We'll then step through each item in _one_ of the lists (doesn't matter which) and make sure they're also present in the other list. If any of them are different, we'll update the categories.

    (Typing this out, it may actually be _technically_ more appropriate and efficient to use `dbApp.categories.some` rather than `!dbApp.categories.every` – although that difference in efficiency won't be noticeable at the scale we're dealing with)

4. Finally, in the case that there are no changes to the categories, the list lengths will be the same, and the second condition will pass as every item in one list is also present in the other list.

Let me know if that helps. I'd be happy to jump on a call and chat through further if it's still unclear, or if I've completely misinterpreted your feedback!

hariombalhara replied:
Ah my bad. Thanks !!

hariombalhara commented on apps/web/pages/api/cron/syncAppMeta.ts:
```diff
@@ -0,0 +1,58 @@
+      updates["categories"] = app.categories as AppCategories[];
+    }
+
+    if (dbApp.dirName !== app.dirName ?? app.slug) {
```
I think it should be like this

```suggestion
    if (dbApp.dirName !== (app.dirName ?? app.slug)) {
```
<img width="479" alt="Screenshot 2023-06-22 at 7 25 50 PM" src="https://github.com/calcom/cal.com/assets/1780212/f0cff9b3-7d91-431d-acf2-4a8ceab7168b">

rjackson replied:
Good catch!

hariombalhara commented on apps/web/pages/api/cron/syncAppMeta.ts:
```diff
@@ -0,0 +1,58 @@
+
+    if (Object.keys(updates).length > 0) {
+      console.log(`🔨 Updating app ${dbApp.slug} with ${Object.keys(updates).join(", ")}`);
+      await prisma.app.update({
+        where: { slug: dbApp.slug },
+        data: updates,
+      });
```
```suggestion
     /* await prisma.app.update({
        where: { slug: dbApp.slug },
        data: updates,
      });*/
```

Because we are doing a one time sync for the first time, there can be unknown issues. I would prefer it to be just a dry run initially and once we review that it's fine, we can do an actual run.

rjackson replied:
Gotcha. Do you think feature flags would be better here? Makes it easier to toggle on / off without having to make code changes.

Would also provide a very quick & convenient way of disabling the sync if it ever does introduce problems.

hariombalhara replied:
Feature flag is more of User facing and this is a dev requirement. This can be an env variable in my opinion, that way it won't affect self-hosters. We are going to verify the query and they might need to do that too if they want to setup that sync.

hariombalhara reviewed (changes requested):
Left 2 more comments.

rjackson committed:
Quick tidy up

rjackson committed:
Merge branch 'main' into recategorise-apps

rjackson committed:
Add dry-run to syncAppMeta via  CRON_ENABLE_APP_SYNC env

rjackson committed:
Replace console.log with logger in syncAppMeta

rjackson commented:
PR updated with an environmental variable, `CRON_ENABLE_APP_SYNC`, to enable/disable the syncAppMeta endpoint. When disabled, the script does a dry-run of the change – reporting what it would do, but not acting on it.

The syncAppMeta endpoint only executes if `CRON_ENABLE_APP_SYNC` is set to `true`. If its omitted, false, or any other value then it runs in the dry-run mode.

I've also swapped over from `console.log` to `@calcom/lib/logger`, using the log prefixes' to indicate what script we're in and whether or not we're in dry-run mode.

**Dry run output**

```txt
@calcom/web:dev: 13:30:08.936 INFO  [api/cron/syncAppMeta] (dry-run) 🧐 Checking DB apps are in-sync with app metadata
@calcom/web:dev: 13:30:08.946 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app daily-video with categories
@calcom/web:dev: 13:30:08.946 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App apple-calendar is up-to-date and correct
@calcom/web:dev: 13:30:08.946 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app routing-forms with categories
@calcom/web:dev: 13:30:08.946 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App caldav-calendar is up-to-date and correct
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App plausible is up-to-date and correct
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App google-calendar is up-to-date and correct
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app whereby with categories
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app google-meet with categories
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app zohocrm with categories
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app jitsi with categories
@calcom/web:dev: 13:30:08.947 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app around with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app wipe-my-cal with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App wordpress is up-to-date and correct
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app huddle01 with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app riverside with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App ga4 is up-to-date and correct
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app typeform with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app webex with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app ping with categories
@calcom/web:dev: 13:30:08.948 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App pipedream is up-to-date and correct
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app campfire with categories
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app raycast with categories
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app sirius_video with categories
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App n8n is up-to-date and correct
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App cron is up-to-date and correct
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App exchange is up-to-date and correct
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app sendgrid with categories
@calcom/web:dev: 13:30:08.949 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App qr_code is up-to-date and correct
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App weather_in_your_calendar is up-to-date and correct
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App gtm is up-to-date and correct
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App fathom is up-to-date and correct
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app closecom with categories
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app whatsapp with categories
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app discord with categories
@calcom/web:dev: 13:30:08.950 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app telegram with categories
@calcom/web:dev: 13:30:08.951 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app signal with categories
@calcom/web:dev: 13:30:08.951 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app sylapsvideo with categories
@calcom/web:dev: 13:30:08.951 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App vimcal is up-to-date and correct
@calcom/web:dev: 13:30:08.951 INFO  [api/cron/syncAppMeta] (dry-run) ✅ App amie is up-to-date and correct
@calcom/web:dev: 13:30:08.951 INFO  [api/cron/syncAppMeta] (dry-run) 🔨 Updating app facetime with categories
```

Live run output:

```txt
@calcom/web:dev: 13:30:47.140 INFO  [api/cron/syncAppMeta] 🧐 Checking DB apps are in-sync with app metadata
@calcom/web:dev: 13:30:47.210 INFO  [api/cron/syncAppMeta] 🔨 Updating app daily-video with categories
@calcom/web:dev: 13:30:47.220 INFO  [api/cron/syncAppMeta] ✅ App apple-calendar is up-to-date and correct
@calcom/web:dev: 13:30:47.220 INFO  [api/cron/syncAppMeta] 🔨 Updating app routing-forms with categories
@calcom/web:dev: 13:30:47.224 INFO  [api/cron/syncAppMeta] ✅ App caldav-calendar is up-to-date and correct
@calcom/web:dev: 13:30:47.225 INFO  [api/cron/syncAppMeta] ✅ App plausible is up-to-date and correct
@calcom/web:dev: 13:30:47.225 INFO  [api/cron/syncAppMeta] ✅ App google-calendar is up-to-date and correct
@calcom/web:dev: 13:30:47.225 INFO  [api/cron/syncAppMeta] 🔨 Updating app whereby with categories
@calcom/web:dev: 13:30:47.230 INFO  [api/cron/syncAppMeta] 🔨 Updating app google-meet with categories
@calcom/web:dev: 13:30:47.234 INFO  [api/cron/syncAppMeta] 🔨 Updating app zohocrm with categories
@calcom/web:dev: 13:30:47.237 INFO  [api/cron/syncAppMeta] 🔨 Updating app jitsi with categories
@calcom/web:dev: 13:30:47.241 INFO  [api/cron/syncAppMeta] 🔨 Updating app around with categories
@calcom/web:dev: 13:30:47.245 INFO  [api/cron/syncAppMeta] 🔨 Updating app wipe-my-cal with categories
@calcom/web:dev: 13:30:47.249 INFO  [api/cron/syncAppMeta] ✅ App wordpress is up-to-date and correct
@calcom/web:dev: 13:30:47.249 INFO  [api/cron/syncAppMeta] 🔨 Updating app huddle01 with categories
@calcom/web:dev: 13:30:47.253 INFO  [api/cron/syncAppMeta] 🔨 Updating app riverside with categories
@calcom/web:dev: 13:30:47.257 INFO  [api/cron/syncAppMeta] ✅ App ga4 is up-to-date and correct
@calcom/web:dev: 13:30:47.258 INFO  [api/cron/syncAppMeta] 🔨 Updating app typeform with categories
@calcom/web:dev: 13:30:47.262 INFO  [api/cron/syncAppMeta] 🔨 Updating app webex with categories
@calcom/web:dev: 13:30:47.266 INFO  [api/cron/syncAppMeta] 🔨 Updating app ping with categories
@calcom/web:dev: 13:30:47.270 INFO  [api/cron/syncAppMeta] ✅ App pipedream is up-to-date and correct
@calcom/web:dev: 13:30:47.270 INFO  [api/cron/syncAppMeta] 🔨 Updating app campfire with categories
@calcom/web:dev: 13:30:47.274 INFO  [api/cron/syncAppMeta] 🔨 Updating app raycast with categories
@calcom/web:dev: 13:30:47.278 INFO  [api/cron/syncAppMeta] 🔨 Updating app sirius_video with categories
@calcom/web:dev: 13:30:47.281 INFO  [api/cron/syncAppMeta] ✅ App n8n is up-to-date and correct
@calcom/web:dev: 13:30:47.281 INFO  [api/cron/syncAppMeta] ✅ App cron is up-to-date and correct
@calcom/web:dev: 13:30:47.281 INFO  [api/cron/syncAppMeta] ✅ App exchange is up-to-date and correct
@calcom/web:dev: 13:30:47.281 INFO  [api/cron/syncAppMeta] 🔨 Updating app sendgrid with categories
@calcom/web:dev: 13:30:47.285 INFO  [api/cron/syncAppMeta] ✅ App qr_code is up-to-date and correct
@calcom/web:dev: 13:30:47.285 INFO  [api/cron/syncAppMeta] ✅ App weather_in_your_calendar is up-to-date and correct
@calcom/web:dev: 13:30:47.285 INFO  [api/cron/syncAppMeta] ✅ App gtm is up-to-date and correct
@calcom/web:dev: 13:30:47.285 INFO  [api/cron/syncAppMeta] ✅ App fathom is up-to-date and correct
@calcom/web:dev: 13:30:47.285 INFO  [api/cron/syncAppMeta] 🔨 Updating app closecom with categories
@calcom/web:dev: 13:30:47.289 INFO  [api/cron/syncAppMeta] 🔨 Updating app whatsapp with categories
@calcom/web:dev: 13:30:47.292 INFO  [api/cron/syncAppMeta] 🔨 Updating app discord with categories
@calcom/web:dev: 13:30:47.296 INFO  [api/cron/syncAppMeta] 🔨 Updating app telegram with categories
@calcom/web:dev: 13:30:47.299 INFO  [api/cron/syncAppMeta] 🔨 Updating app signal with categories
@calcom/web:dev: 13:30:47.302 INFO  [api/cron/syncAppMeta] 🔨 Updating app sylapsvideo with categories
@calcom/web:dev: 13:30:47.306 INFO  [api/cron/syncAppMeta] ✅ App vimcal is up-to-date and correct
@calcom/web:dev: 13:30:47.306 INFO  [api/cron/syncAppMeta] ✅ App amie is up-to-date and correct
@calcom/web:dev: 13:30:47.307 INFO  [api/cron/syncAppMeta] 🔨 Updating app facetime with categories
```

hariombalhara reviewed (approved):
LGTM now !! Thank you so much @rjackson. I learned a few things from you while collaborating with you on this 🙏

hariombalhara commented:
@keithwillcode This should be tested on staging. Maybe @ericrommel can do this?

rjackson commented:
> LGTM now !! Thank you so much @rjackson. I learned a few things from you while collaborating with you on this 🙏

Yeah no worries!

It's been a really fun & interesting task 😆 Sounded like a small job in the first place, but exposed a few ~~warts~~ opportunities for improvement behind-the-scenes. Really useful to get a better understanding of the app store & related architecture as well.

Thanks for your feedback and support throughout  🎉

PeerRich commented:
/tip 50

PeerRich commented:
tippin a fiffy @rjackson
